greeting.py

```diff
@@ -1,13 +1,42 @@
 def greet(name):
     if name is None:
-        name = "my friend"
-    elif isinstance(name, list):
-        name = f"{name[0]} and {name[1]}"
+        return get_greet_from_string("my friend")
+    if isinstance(name, list):
+        return get_greet_from_list(name)
     else:
-        name = str(name)
+        return get_greet_from_string(str(name))
+
 
+def get_greet_from_string(name):
     if name.isupper():
-        result = f"HELLO {name}!"
+        return f"HELLO {name}!"
+    else:
+        return f"Hello, {name}."
+
+
+def get_greet_from_list(name):
+    normal_name = [n for n in name if not n.isupper()]
+    shouting_name = [n for n in name if n.isupper()]
+    normal_greet = get_greet_string_from_list(normal_name, is_shouting=False) if normal_name else ""
+    shouting_greet = get_greet_string_from_list(shouting_name, is_shouting=True) if shouting_name else ""
+
+    if normal_name and shouting_name:
+        return f"{normal_greet} AND {shouting_greet}"
+    elif normal_name or shouting_name:
+        return normal_greet + shouting_greet
     else:
-        result = f"Hello, {name}."
-    return result
+        return greet(None)
+
+
+def get_greet_string_from_list(name, **kwargs):
```

Comment:
here you still can specify is_shouting by name, it's fine. maybe event do
```suggestion
def get_greet_string_from_list(name, *, is_shouting):
```
this means that is_shouting can only be passed as `kwarg`, by name

Reply:
https://www.python.org/dev/peps/pep-3102/